### PR TITLE
Finish basic support for enumerations

### PR DESCRIPTION
**Pull request by davisp**
This finishes out the basic support for adding enumeration support to arrays and schemas.

### CODE REVIEW

File: tiledb/api/src/array/attribute/mod.rs [hunk 1]
```diff
@@ -345,6 +365,25 @@ impl Builder {
         Ok(self)
     }
 
+    /// Set the name of the enumeration to use.
+    ///
+    /// Note that when building schemas, the enumeration must have been added
+    /// to the schema for adding an attribute that references it.
```
Comment (rroelke) [inline tiledb/api/src/array/attribute/mod.rs:371]: Does failing to abide by this result in an error here?  Or when the attribute is added to the schema? Or at schema creation time?

File: tiledb/utils/src/option.rs [hunk 2]
```diff
@@ -114,6 +114,24 @@ where
     }
 }
 
+impl<T> OptionSubset for Box<T>
```
Comment (rroelke) [inline tiledb/utils/src/option.rs:117]: This probably oughta be more general, e.g.
`impl<T> OptionSubset for &T where T: OptionSubset`

I just had a look at `PartialEq` and it looks like it has both this version and also the one you added.  So, up to you if you want to add it.

File: tiledb/api/src/array/mod.rs [hunk 3]
```diff
@@ -1238,4 +1266,79 @@ pub mod tests {
...
+            .build();
+
+            DomainBuilder::new(&ctx)?.add_dimension(dim)?.build()
+        };
+
+        let enmr = EnumerationBuilder::new(
+            &ctx,
+            "flintstones",
+            Datatype::StringUtf8,
+            "fredwilmageorgebetty".as_bytes(),
```
Comment (rroelke) [inline tiledb/api/src/array/mod.rs:1296]: It's certainly not a request for changes in this PR but it feels a little wrong that the default cell val num is single but the other arguments here expect something else.

If offsets is `Some` then it should probably infer CellValNum::Var

**Review by rroelke — APPROVED** [review]
I'll leave a weak suggestion to add a test that the enumeration values which we put *in* to the schema are the same as those that we read *out*.  Otherwise this looks good

File: tiledb/api/src/array/attribute/mod.rs (hunk 1 above)
Comment (davisp) [reply tiledb/api/src/array/attribute/mod.rs:371]: You'll get an error from TileDB core yeah: https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/array_schema/array_schema.cc#L810-L820

File: tiledb/api/src/array/mod.rs (hunk 3 above)
Comment (davisp) [reply tiledb/api/src/array/mod.rs:1296]: I'll add a story. It looks like Enumeration predates the CellValNum enum so it'll probably spider in the tests a bit. Also has the feel for another good "first story" for onboarding.

Comment (davisp) [reply tiledb/api/src/array/mod.rs:1296]: https://app.shortcut.com/tiledb-inc/story/54981

File: tiledb/utils/src/option.rs (hunk 2 above)
Comment (davisp) [reply tiledb/utils/src/option.rs:117]: Huh. I'll poke at that some more. I gave up a bit here when the first box wasn't doing exactly what I wanted and just added the `Box<[T]>` variant. Assuming your approach works I'll totes switch that out.

Comment (davisp) [reply tiledb/utils/src/option.rs:117]: I attempted using this approach and couldn't quite figure out how to make it work correctly. The `impl` you suggested gets into a recursive loop that causes a stack overflow. Reading up more, I also found this post that seems to be related which just says that it has to do with fundamental types, which Option isn't.

**Pull request merged**